packages/astro/src/runtime/server/render/component.ts

```diff
@@ -515,7 +515,7 @@ export async function renderComponentToString(
 	// Handle head injection if required. Note that this needs to run early so
 	// we can ensure getting a value for `head`.
 	let head = '';
-	if (nonAstroPageNeedsHeadInjection(Component)) {
+	if (isPage && !result.partial && nonAstroPageNeedsHeadInjection(Component)) {
```

Comment:
We now only try to render the head if this component is a page and is non-partial. Otherwise there's chance we could render a component but not use it (with how the new `renderJSX` flow works)